[GraphQL] Check listing visibility against disco server

franckc opened the pull request:
# Description:

Address issue #3110 , except for the unpinning part which will be addressed in a separate PR.

In performance mode (aka centralized), added code in the graphql server to check against the discovery server if a listing is visible before returning it. Since the listing visibility might be sensitive data, access to that field on the discovery server is protected via a header based authentication. We'll have the graphql server setting that header when calling the discovery server.

Also did a small refactoring by adding the networkId definition to the config in order to clean up listingId construction in the graphQL code I modified.

### Checklist:

- [X] Test your work and double-check to confirm that you didn't break anything
- [ ] [Wrap any new text/strings with `fbt`](https://github.com/OriginProtocol/origin/tree/master/dapps/marketplace/translation#wrapping-text) for translation
- [ ] If there are any new/changed strings to translate, [`npm run translate`](https://github.com/OriginProtocol/origin/tree/master/dapps/marketplace/translation#integrating-translations)
- [ ] Update any relevant READMEs and [docs](https://github.com/OriginProtocol/origin/tree/master/docs)

sparrowDom commented on packages/graphql/src/resolvers/marketplace/listings.js:
```diff
@@ -79,12 +80,20 @@ async function resultsFromIds({ after, ids, first, totalCount, fields }) {
+      ids.map(async id => {
+        // If a discovery server is configured, check the listing's visibility.
+        if (contracts.discovery) {
+          if (!(await listingIsVisible(id))) {
```
I would also vote for making a single request and not affecting the pagination. Though the fix might be more involved...

In graphQL schema we support the [listingIds](https://github.com/OriginProtocol/origin/blob/master/packages/graphql/src/typeDefs/Marketplace.js#L149) property but I don't see discovery [implementing it](https://github.com/OriginProtocol/origin/blob/master/infra/discovery/src/apollo/resolvers.js#L13-L36). So some change to discovery server would be required

franckc commented on infra/discovery/src/apollo/resolvers.js:
```diff
@@ -10,27 +12,46 @@ const web3 = new Web3(process.env.PROVIDER_URL || 'http://localhost:8545')
-        stats: {
-          maxPrice: stats.maxPrice,
-          minPrice: stats.minPrice
+      if (args.ids) {
```
Note: this code path is not used in this PR. I started adding it and then decided it was better to add a different query (visibleListingIds). But since it could be handy to query listings from discovery by ids, I'm leaving it in here for future use...

franckc commented:
Integrated feedback from previous round of review:
 - Improved pagination by filtering batch of ids rather than one at a time
 - Added a new query to the discovery GraphQL server called visibleListingIds. It filters out non-visible id and gets called by the graphQL layer if discovery is enabled. Note that I originally started doing the visibility check by querying the "listings" end resolver of the discovery server. But this causes several issues in the listener and the DApp due to timing. The listings endpoint relies on the listing data being indexed, which is not the case if called by the listener prior to indexing the listing or by the DApp shortly after a new listing is created. Therefore, rather than relying on indexed listing data, I decided to rely on the visibleListingIds which calculates a listing ids blacklist based on the moderator data stored in the discovery_tag_action table.
 - visibleListingIds filters but does not disclose the exact tags that are set on listings. So I think it's ok to leave this endpoint public with no authentication. The same information could anyway be obtained by running searches against the discovery server and seeing which listings are filtered out.
 - Added a TODO in the code to add the ability for the DApp to decide whether the visibility check should be performed at the graphQL layer or not. We can implement this in the future if our volume grows. It could be used to solve the edge case of wanting able to show a hidden listing to the buyer/seller if there is an in-flight offer on it. I don't think it's worth adding complexity to the code now given we have a tiny volume of offers.

nick commented:
LGTM 👍

DanielVF commented:
Looks good!
